### ISSUE TITLE
etag + [ if-none-match ] issues

### ISSUE THREAD

**Issue by foxontherock** [basecamp/bc3-api#177]
Hi, we are developing an app using the Basecamp 3 API.

We keep some data in "local memory caching", like list of all users, projects, lists, list-groups... 
And we plan to use the ETag / If-None-Match method, to refresh our cache when needed.

But, it doesn't seem to work well.

Example
I call "/projects.json", I get a new ETag.
And, on my following calls, I send it as If-None-Match header.
Sometimes, I get 304.  But not always.
Even if my data hasn't change.

The proof is, that if I re-do the same query, (with the initial etag, not the new one), then I may get 304 again.  So, the 200 I get previously is a bad one, I should have get another 304.

Sometimes, I get "false positive" 200 response with data, instead of 304.

And I get them enough to consider not using it.

Also, are these "304" requests supposed to be faster?

Yes I get "no data" as results, but the wait time is almost the same!  Where's the advantage?  And that 304 call affects my 50 per 10 seconds call limit.  

So, it's the same to get real data or 304 response, and I see no advantage on time and call rate limit.

Thank you

**Comment by rosa** [comment]
Hey @foxontherock! First of all, sorry for the delay. I saw you contacted us via Support shortly afterwards and followed the conversation over there, and then I forgot to reply here. 

It looks like we have some kind of non-determinism when computing the ETag in that particular endpoint (it doesn't happen in other endpoints like for example [getting a comment](https://github.com/basecamp/bc3-api/blob/master/sections/comments.md#get-a-comment)). I could reproduce the same issue you describe in `/projects.json` and also in `/projects/:project_id.json`. I haven't figured out the cause yet, but I'll keep digging.

**Comment by doudou** [comment]
I'm having an issue that looks alike with the `recordings.json` endpoint. I get a 200 regardless of whether I provide `If-None-Match` or `If-Modified-Since`.

Not sure whether I should create a new issue for this, or leave it here.

**Comment by rosa** [comment]
@doudou thanks for letting us know! Here it's fine, we haven't been able to pinpoint the cause of this problem yet, but I'll update here as soon as we do.

**Comment by doudou** [comment]
> Here it's fine

Mmmm ... I'll quadruple-check on my side. I'm probably doing something wrong.

**Comment by doudou** [comment]
... I was, it seems. Not sure what. I did a small one/off test just to check, but managed to mess it up. With a complete clean implementation, it works. Apologies for the noise.

**Comment by rosa** [comment]
@doudou no worries, and thank you for letting us know!

**Comment by rosa** [comment]
I'm closing this issue for now. We've spent more time looking into but haven't been able to pinpoint the cause of the non-determinism when computing the ETag in that endpoint, unfortunately. It's unlikely we spend more time investigating since it's a very minor issue and it might not be a huge advantage to rely on 304 responses there as you said (because you can't avoid most of the time spent on the network and because the projects endpoints have a large cache churn anyway).

**Comment by ericdude4** [comment]
Also being blocked on this issue. I guess as long as you are okay with my API polling your API to check for new projects (Which I then create web hook subscriptions on) on regular intervals, then it's not a problem.
